### ISSUE TITLE
extension does not take np.ndarray argument

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/pynwb#333]
```python
import numpy as np
import random

from pynwb.spec import NWBDatasetSpec, NWBNamespaceBuilder, NWBGroupSpec
from pynwb import get_class, load_namespaces, NWBFile, NWBHDF5IO

ns_path = "spikey.namespace.yaml"
ext_source = "spikey.extensions.yaml"
PopulationSpikeTimes = NWBGroupSpec(doc='Population Spike Times',
                                      datasets=[
                                          NWBDatasetSpec(doc='spike times for all neurons', 
                                                         shape=(1,), 
                                                         name='times', dtype='float'),
                                          NWBDatasetSpec(doc='data pointer', 
                                                         shape=(1,),
                                                         name='data_pointer', dtype='int')
                             ],
                           neurodata_type_def='PopulationSpikeTimes',
                           neurodata_type_inc='NWBContainer')
ns_builder = NWBNamespaceBuilder('spikey extensions', "spikey")
ns_builder.add_spec(ext_source, PopulationSpikeTimes)
ns_builder.export(ns_path)

# generate fake data
duration = 10
nspikes = 1000
nunits = 5

times = np.random.rand(nspikes) * duration
data_pointer = np.sort(random.sample(range(len(times)), nunits))

## pull types from yaml files
load_namespaces(ns_path)
PopulationSpikeTimes = get_class('PopulationSpikeTimes', 'spikey')

pop_data = PopulationSpikeTimes(name='example population', source='source',
                                                          times=times, data_pointer=data_pointer)
```
gives the following error:

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-46-eb8a530669b8> in <module>()
      3 PopulationSpikeTimes = get_class('PopulationSpikeTimes', 'spikey')
      4 
----> 5 pop_data = PopulationSpikeTimes(name='example population', source='source', times=times, data_pointer=data_pointer)
      6 
      7 

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    303                 if parse_err:
    304                     msg = ', '.join(parse_err)
--> 305                     raise_from(TypeError(msg), None)
    306                 return func(self, **parsed['args'])
    307         else:

~/anaconda3/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

TypeError: incorrect type for 'times' (got 'ndarray', expected 'list, tuple, dict or set'), incorrect type for 'data_pointer' (got 'ndarray', expected 'list, tuple, dict or set')
```
I can convert the ndarray into a list in this case and avoid the error, but I feel like this extension should be able to accept a list
### Environment
Please describe your environment according to the following bullet points.
- **Python Executable:** Conda
- **Python Version:** Python 3.6
- **Operating System:** macOS
- **Pynwb Version:** 0.2.0

**Comment by bendichter** [comment]
also,

```python
import numpy as np
import random

from pynwb.spec import NWBDatasetSpec, NWBNamespaceBuilder, NWBGroupSpec
from pynwb import get_class, load_namespaces, NWBFile, NWBHDF5IO

from datetime import datetime

ns_path = "spikey.namespace.yaml"
ext_source = "spikey.extensions.yaml"
PopulationSpikeTimes = NWBGroupSpec(doc='Population Spike Times',
                                      datasets=[
                                          NWBDatasetSpec(doc='spike times for all neurons', 
                                                         shape=(None,1), 
                                                         name='times', dtype='float'),
                                          NWBDatasetSpec(doc='data pointer', 
                                                         shape=(None,1),
                                                         name='data_pointer', dype='int')
                             ],
                           neurodata_type_def='PopulationSpikeTimes',
                           neurodata_type_inc='NWBContainer')
ns_builder = NWBNamespaceBuilder('spikey extensions', "spikey")
ns_builder.add_spec(ext_source, PopulationSpikeTimes)
ns_builder.export(ns_path)

# generate fake data
duration = 10
nspikes = 1000
nunits = 5

times = np.random.rand(nspikes) * duration
data_pointer = np.sort(random.sample(range(len(times)), nunits))

## pull types from yaml files
load_namespaces(ns_path)
PopulationSpikeTimes = get_class('PopulationSpikeTimes', 'spikey')

pop_data = PopulationSpikeTimes(name='example population', source='source',
                                times=list(times), data_pointer=list(data_pointer))


## write to file
f = NWBFile(file_name='tmp.nwb',
            source='me',
            session_description='my first synthetic recording',
            identifier='EXAMPLE_ID',
            session_start_time=datetime.now(),
            experimenter='Dr. Bilbo Baggins',
            lab='Bag End Labatory',
            institution='University of Middle Earth at the Shire',
            experiment_description='empty',
            session_id='LONELYMTN')

population_module = f.create_processing_module(name='0', source='source',
                                               description='description') 
population_module.add_container(pop_data)


io = NWBHDF5IO('tmp.nwb', mode='w')       # A little less work on your part
io.write(f)
io.close()
```

gives error

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-73-b04de35240c1> in <module>()
     58 
     59 io = NWBHDF5IO('tmp.nwb', mode='w')       # A little less work on your part
---> 60 io.write(f)
     61 io.close()

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/backends/io.py in write(self, **kwargs)
     31     def write(self, **kwargs):
     32         container = getargs('container', kwargs)
---> 33         f_builder = self.__manager.build(container, source=self.__source)
     34         self.write_builder(f_builder)
     35 

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
     31         result = self.__builders.get(container_id)
     32         if result is None:
---> 33             result = self.__type_map.build(container, self, source=getargs('source', kwargs))
     34             self.prebuilt(container, result)
     35         return result

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
    984             raise ValueError('No ObjectMapper found for container of type %s' % str(container.__class__.__name__))
    985         else:
--> 986             builder = attr_map.build(container, manager, source=getargs('source', kwargs))
    987         namespace, data_type = self.__get_container_ns_dt(container)
    988         builder.set_attribute('namespace', namespace)

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
    436             builder = GroupBuilder(name, parent=parent, source=source)
    437             self.__add_datasets(builder, self.__spec.datasets, container, manager, source)
--> 438             self.__add_groups(builder, self.__spec.groups, container, manager, source)
    439             self.__add_links(builder, self.__spec.links, container, manager, source)
    440         else:

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_groups(self, builder, groups, container, build_manager, source)
    519                             if isinstance(item, Container):
    520                                 self.__add_containers(sub_builder, spec, item, build_manager, source)
--> 521                 self.__add_groups(sub_builder, spec.groups, container, build_manager, source)
    522                 empty = sub_builder.is_empty()
    523                 if not empty or (empty and isinstance(spec.quantity, int)):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_groups(self, builder, groups, container, build_manager, source)
    534                     attr_value = getattr(container, attr_name, None)
    535                     if attr_value is not None:
--> 536                         self.__add_containers(builder, spec, attr_value, build_manager, source)
    537 
    538     def __add_containers(self, builder, spec, value, build_manager, source):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_containers(self, builder, spec, value, build_manager, source)
    559                 raise ValueError(msg % value.__class__.__name__)
    560             for container in values:
--> 561                 self.__add_containers(builder, spec, container, build_manager, source)
    562 
    563     def __get_subspec_values(self, builder, spec, manager):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_containers(self, builder, spec, value, build_manager, source)
    538     def __add_containers(self, builder, spec, value, build_manager, source):
    539         if isinstance(value, Container):
--> 540             rendered_obj = build_manager.build(value, source=source)
    541             # use spec to determine what kind of HDF5
    542             # object this Container corresponds to

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
     31         result = self.__builders.get(container_id)
     32         if result is None:
---> 33             result = self.__type_map.build(container, self, source=getargs('source', kwargs))
     34             self.prebuilt(container, result)
     35         return result

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
    984             raise ValueError('No ObjectMapper found for container of type %s' % str(container.__class__.__name__))
    985         else:
--> 986             builder = attr_map.build(container, manager, source=getargs('source', kwargs))
    987         namespace, data_type = self.__get_container_ns_dt(container)
    988         builder.set_attribute('namespace', namespace)

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
    436             builder = GroupBuilder(name, parent=parent, source=source)
    437             self.__add_datasets(builder, self.__spec.datasets, container, manager, source)
--> 438             self.__add_groups(builder, self.__spec.groups, container, manager, source)
    439             self.__add_links(builder, self.__spec.links, container, manager, source)
    440         else:

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_groups(self, builder, groups, container, build_manager, source)
    534                     attr_value = getattr(container, attr_name, None)
    535                     if attr_value is not None:
--> 536                         self.__add_containers(builder, spec, attr_value, build_manager, source)
    537 
    538     def __add_containers(self, builder, spec, value, build_manager, source):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_containers(self, builder, spec, value, build_manager, source)
    559                 raise ValueError(msg % value.__class__.__name__)
    560             for container in values:
--> 561                 self.__add_containers(builder, spec, container, build_manager, source)
    562 
    563     def __get_subspec_values(self, builder, spec, manager):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_containers(self, builder, spec, value, build_manager, source)
    538     def __add_containers(self, builder, spec, value, build_manager, source):
    539         if isinstance(value, Container):
--> 540             rendered_obj = build_manager.build(value, source=source)
    541             # use spec to determine what kind of HDF5
    542             # object this Container corresponds to

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
     31         result = self.__builders.get(container_id)
     32         if result is None:
---> 33             result = self.__type_map.build(container, self, source=getargs('source', kwargs))
     34             self.prebuilt(container, result)
     35         return result

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
    984             raise ValueError('No ObjectMapper found for container of type %s' % str(container.__class__.__name__))
    985         else:
--> 986             builder = attr_map.build(container, manager, source=getargs('source', kwargs))
    987         namespace, data_type = self.__get_container_ns_dt(container)
    988         builder.set_attribute('namespace', namespace)

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in build(self, **kwargs)
    435         if isinstance(self.__spec, GroupSpec):
    436             builder = GroupBuilder(name, parent=parent, source=source)
--> 437             self.__add_datasets(builder, self.__spec.datasets, container, manager, source)
    438             self.__add_groups(builder, self.__spec.groups, container, manager, source)
    439             self.__add_links(builder, self.__spec.links, container, manager, source)

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __add_datasets(self, builder, datasets, container, build_manager, source)
    486     def __add_datasets(self, builder, datasets, container, build_manager, source):
    487         for spec in datasets:
--> 488             attr_value = self.get_attr_value(spec, container)
    489             # TODO: add check for required datasets
    490             if attr_value is None:

~/anaconda3/lib/python3.6/site-packages/pynwb/form/utils.py in func_call(*args, **kwargs)
    304                     msg = ', '.join(parse_err)
    305                     raise_from(TypeError(msg), None)
--> 306                 return func(self, **parsed['args'])
    307         else:
    308             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in get_attr_value(self, **kwargs)
    387             return None
    388         else:
--> 389             return self.__convert_value(attr_val, spec)
    390 
    391     def __convert_value(self, value, spec):

~/anaconda3/lib/python3.6/site-packages/pynwb/form/build/map.py in __convert_value(self, value, spec)
    402                 ret = value
    403             else:
--> 404                 if 'text' in spec.dtype:
    405                     if spec.dims is not None:
    406                         ret = list(map(str, value))

TypeError: argument of type 'NoneType' is not iterable
```

**Comment by neuromusic** [comment]
hey @nicain is this similar to your population spike train extension?

**Comment by nicain** [comment]
Could be: https://github.com/nicain/population_spike_train/blob/master/generate_extension.py

CCing for visibility @V0L1T10N @CellAssembly @kaeldai

@bendichter The above are Sergey, Kael, and Yazan (not necessarily in that order...) from Anton's group.

**Comment by bendichter** [comment]
Thanks for showing me this @nicain. Does it store spike times in order in `timestamps` and gid in `data` of `TimeSeries`? Is the purpose of the extension just to constrain `data` to take only `int`s for the gid? If so, it seems very similar to the `Clustering` object, which isn't a `TimeSeries` but I think should be and I'm working with Andrew to change it

**Comment by ajtritt** [comment]
should have closed with #334, but there was a typo in the PR message
